firmware/library/L2_Utilities/utility.cpp

```diff
@@ -0,0 +1,94 @@
+#include <cstdlib>
+
+#include "utility.hpp"
+
+uint32_t utility::Random(uint32_t min, uint32_t max)
+{
+    return rand() % (max - min + 1) + min;
+}
+void utility::SetRandomSeed(uint32_t seed)
+{
+    srand(seed);
+}
+uint32_t utility::BitExtract(uint32_t target, uint8_t start_bit, uint8_t width)
+{
+    uint32_t mask = 0;
+    for (int i = 0; i < width; i++)
+    {
+        mask = (mask << 1) + 1;
+    }
+    uint8_t true_start_bit = static_cast<uint8_t>(start_bit - (width - 1));
+    mask <<= true_start_bit;
+    return (target & mask) >> true_start_bit;
+}
+uint32_t utility::BitPlace(uint32_t target, uint8_t start_bit,
+                           uint32_t new_bits, uint8_t width)
+{
+    uint32_t temp = target;
+    for (uint8_t i = 0; i < width; i++)
+    {
+        temp = utility::BitClear(target, uint8_t(start_bit - i));
+    }
+    uint32_t mask = new_bits;
+    mask <<= (start_bit - (width - 1));
+    return temp | mask;
+}
+
+uint32_t utility::BitSet(uint32_t target, uint8_t bit_position)
+{
+    uint32_t mask = 1 << bit_position;
+    return target | mask;
+}
+uint32_t utility::BitClear(uint32_t target, uint8_t bit_position)
+{
+    uint32_t mask = 1 << bit_position;
+    mask          = ~mask;
+    return target & mask;
+}
+uint32_t utility::BitRead(uint32_t target, uint8_t bit_position)
+{
+    uint32_t mask       = 1 << bit_position;
+    uint32_t masked_var = target & mask;
+    return masked_var >> bit_position;
+}
+uint32_t utility::Min(uint32_t compare_1, uint32_t compare_2)
+{
+    if (compare_1 >= compare_2){
+        return compare_2;
+    }
+    else{
+        return compare_1;
+    }
+}
+uint32_t utility::Max(uint32_t compare_1, uint32_t compare_2)
+{
+    if (compare_2 >= compare_1){
+        return compare_2;
+    }
+    else{
+        return compare_1;
+    }
+}
+uint32_t utility::Map(uint32_t value, uint32_t current_min,
+                      uint32_t current_max, uint32_t new_min, uint32_t new_max)
+{
+    int64_t map_value       = value;
+    int64_t map_current_min = current_min;
+    int64_t map_current_max = current_max;
+    int64_t map_new_min     = new_min;
+    int64_t map_new_max     = new_max;
+    return static_cast<uint32_t>((map_value - map_current_min) *
```

Comment:
There is an issue with using ints and implicitly casting them into floats and back to ints. There may be truncation through out the operations which can lead to improper values. 

The best course of action is to make an overloaded version of this function that takes floats and inputs and outputs a float. The version of map that uses ints would handle the conversion to floats and simply call the float version.

Within the floating point version you should convert everything to doubles (use doubles rather than floats to improve precision during mathematical operations) and cast down to float at the end (reduce size at the end).

Reply:
Can you also template this function as well. With appropriate testes for int, float and double.